src/hotspot/share/runtime/os.cpp

```diff
@@ -961,10 +962,14 @@ static void print_hex_location(outputStream* st, address p, int unitsize) {
 #endif // 32-bit, qwords
   intptr_t i = 0;
   if (read_safely_from((intptr_t*)pa, &i)) {
+    // bytes:   CA FE BA BE DE AD C0 DE
+    // bytoff:   0  1  2  3  4  5  6  7
+    // LE bits  0  8  16 24 32 40 48 56
+    // BE bits  56 48 40 32 24 16  8  0
     const int offset = (int)(p - (address)pa);
     const int bitoffset =
       LITTLE_ENDIAN_ONLY(offset * BitsPerByte)
-      BIG_ENDIAN_ONLY((int)(sizeof(intptr_t) - 1 - offset) * BitsPerByte);
+      BIG_ENDIAN_ONLY((int)((sizeof(intptr_t) - unitsize - offset) * BitsPerByte));
```

Comment:
Um, my head hurts a bit thinking about this. So for 64-bit `sizeof(intptr_t)` = 8 and unitsize=8, `bitoffset` is negative? Which we then feed into `bitfield`, which does the shift by the negative value, which is undefined behavior?

Reply:
offset is based on the print address, and that one is aligned to unitsize:

https://github.com/openjdk/jdk/blob/caef3d5ac2ba34610b2fcd65045d9c9185c5de06/src/hotspot/share/runtime/os.cpp#L995

which I now also assert at the entry of print_hex_location.

For unitsize = 8, on 64-bit, print address should equal the address of the intptr_t, and the offset should be 0.

Hence also the align_as in the test: I want to make sure the address range is always aligned in a way that does not conflict with unitsize alignment.

Reply:
Mhm. What worries me is that this is a common part of error handling code. `assert` does not really help us here, because callers can still hand us unaligned pointers, especially in release bits, which would face the error in this code? In other words, this code should be bullet-proof, and not rely on any externality.

One could argue that `print_hex_location` is an internal function that is protected by what `os::print_hex_dump` does. But then aligning `char buf[256]` prior to call to `os::print_hex_dump` in the test is actually a red flag! `os::print_hex_dump` should be immune to misaligned `buf`!

Reply:
> Mhm. What worries me is that this is a common part of error handling code. assert does not really help us here, because callers can still hand us unaligned pointers, especially in release bits, which would face the error in this code? 

We always align down the pointer by unit size: https://github.com/openjdk/jdk/blob/caef3d5ac2ba34610b2fcd65045d9c9185c5de06/src/hotspot/share/runtime/os.cpp#L995

The assert in print_hex_location (a local static function, not externally exposed) just asserts this. More for documentation purposes really.

> But then aligning char buf[256] prior to call to os::print_hex_dump in the test is actually a red flag! 

When I wrote the test originally, I allocated the input buffer with malloc() in order to have a guaranteed alignment. But I added explicit tests for crooked aligned pointers: 

https://github.com/openjdk/jdk/blame/caef3d5ac2ba34610b2fcd65045d9c9185c5de06/test/hotspot/gtest/runtime/test_os.cpp#L237

I'm nothing if not thorough :)

About the alignas, that was a stupid thinking error for multiple reasons and I'll remove it again (its completely pointless).